target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts

```diff
@@ -0,0 +1,415 @@
+			};
+			partition@1340000 {
+				label = "extra";
+				reg = <0x1340000 0x4000000>;
+			};
+		};
+	};
+};
+
+&mdio0 {
+	status = "okay";
+
+	pinctrl-0 = <&mdio0_pins>;
+	pinctrl-names = "default";
+
+	ethernet-phy@0 {
+		reg = <0x0>;
+		qca,ar8327-initvals = <
+			0x00004 0x7600000   /* PAD0_MODE */
+			0x00008 0x1000000   /* PAD5_MODE */
+			0x0000c 0x80        /* PAD6_MODE */
+			0x000e4 0xaa545     /* MAC_POWER_SEL */
+			0x000e0 0xc74164de  /* SGMII_CTRL */
+			0x0007c 0x4e        /* PORT0_STATUS */
+			0x00094 0x4e        /* PORT6_STATUS */
+			>;
+	};
+
+	phy7: ethernet-phy@7 {
+		reg = <7>;
+	};
+};
+
+&gmac0 {
+	status = "okay";
+	phy-mode = "rgmii";
+	qcom,id = <0>;
+
+	nvmem-cells = <&fw_env_macaddr_4_unknown>;
+	nvmem-cell-names = "mac-address";
+
+	pinctrl-0 = <&rgmii2_pins>;
+	pinctrl-names = "default";
+
+	fixed-link {
+		speed = <1000>;
+		full-duplex;
+	};
+};
+
+&gmac1 {
+	status = "okay";
+	phy-mode = "sgmii";
+	qcom,id = <1>;
+
+	nvmem-cells = <&fw_env_macaddr_0_eth1_lan>;
+	nvmem-cell-names = "mac-address";
+
+	fixed-link {
+		speed = <1000>;
+		full-duplex;
+	};
+};
+
+&gmac3 {
+	status = "okay";
+	phy-mode = "sgmii";
+	qcom,id = <3>;
+	phy-handle = <&phy7>;
+
+	nvmem-cells = <&fw_env_macaddr_1_eth2_wan>;
+	nvmem-cell-names = "mac-address";
+};
+
+&adm_dma {
+	status = "okay";
+};
+
+&usb3_1 {
+	status = "okay";
+};
+
+&pcie0 {
+	status = "okay";
+	reset-gpio = <&qcom_pinmux 3 GPIO_ACTIVE_HIGH>;
+	pinctrl-0 = <&pcie0_pins>;
+	pinctrl-names = "default";
+
+	bridge@0,0 {
+		reg = <0x00000000 0 0 0 0>;
+		#address-cells = <3>;
+		#size-cells = <2>;
+		ranges;
+
+		wifi0: wifi@1,0 {
+			compatible = "pci168c,0046";
+			reg = <0x00010000 0 0 0 0>;
+
+			nvmem-cells = <&ART_precal_0_ath0_5G>, <&fw_env_macaddr_3_ath0_5G>;
+			nvmem-cell-names = "pre-calibration", "mac-address";
+		};
+	};
+};
+
+&pcie1 {
+	status = "okay";
+	reset-gpio = <&qcom_pinmux 48 GPIO_ACTIVE_HIGH>;
+	pinctrl-0 = <&pcie1_pins>;
+	pinctrl-names = "default";
+	max-link-speed = <1>;
+
+	bridge@0,0 {
+		reg = <0x00000000 0 0 0 0>;
+		#address-cells = <3>;
+		#size-cells = <2>;
+		ranges;
+
+		wifi1: wifi@1,0 {
+			compatible = "pci168c,0040";
+			reg = <0x00010000 0 0 0 0>;
+
+			nvmem-cells = <&ART_precal_1_ath1_2G>, <&fw_env_macaddr_2_ath1_2G>;
+			nvmem-cell-names = "pre-calibration", "mac-address";
+		};
+	};
+};
+
+&ART {
+	compatible = "nvmem-cells";
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	ART_precal_0_ath0_5G: precal@1000 { reg = <0x1000 0x2f20>; };
+	ART_precal_1_ath1_2G: precal@5000 { reg = <0x5000 0x2f20>; };
+};
+
+&fw_env {
+	compatible = "nvmem-cells";
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	fw_env_macaddr_0_eth1_lan: macaddr@0  { reg = <0x00 0x6>; };
+	fw_env_macaddr_1_eth2_wan: macaddr@6  { reg = <0x06 0x6>; };
+	fw_env_macaddr_2_ath1_2G:  macaddr@c  { reg = <0x0c 0x6>; };
+	fw_env_macaddr_3_ath0_5G:  macaddr@12 { reg = <0x12 0x6>; };
+	fw_env_macaddr_4_unknown:  macaddr@18 { reg = <0x18 0x6>; };
+};
```

Comment:
One last comment, could you move this into the node definition? And then please remove the partition node labels (`ART:` and `fw_env:`).
https://github.com/openwrt/openwrt/blob/c0a6fefd9b74af03a30810be20f0ce5504447b3e/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts#L208-L231

While at it, could you consider following the convention for the nvmem-cells nodes?
```
precal_ART_1000: precal@1000 {
	reg = <0x1000 0x2f20>;
};

macaddr_fw_env_18: macaddr@18 {
	reg = <0x18 0x6>;
};
```

I would be happy if you add a small note for the MAC addresses in the commit message like 70c12d26ca6e.

Reply:
@mans0n,
FYI: not much to say about the MACs, except that i haven't seen `fw_env_macaddr_4_unknown` being used by stock fw. stock doesn't set the addresses of interfaces before enabling them, and i've never seen eth0 enabled. i still chose to give it an address over leaving it unconfigured; and since `fw_env_macaddr_4_unknown` is provisioned (the 5 MACs are sequential) and would otherwise go unused, i gave it that.

Reply:
I think it's okay, please just write down about your decision in the commit message.

Reply:
hi @mans0n 

let me first say that i really appreciate the time you take to review and help with PRs.

unfortunately your review came over a week past my last commit, and by that time i had put the router back together and commissioned it. it's not that i can't grab it back; i can, and i would if there were correctness issues, but i can't justify the time potentially needed -especially if i required serial again- for changes that are style-only.

btw i favor my style, where variables are named according to their usage/content rather than their addresses, negating the need for documenting variable contents in the commit message. i prefer array indices over addresses, and separating concerns so that i don't have to go over the entire partition table to find information that -other than being stored in flash- has not relation to the partition table. but of course style is a matter of opinion entirely.

your criticism while working on this was very much appreciated. it forced me to dive in and learn stuff i had no idea of, and it was fun to do that, which is my real motivation for collaborating in free software. but right now i have very little time and i want to invest it where returns are higher: providing a safer upgrade mechanism for openwrt -as i said i would 2/3 weeks ago. and i haven't even started because of lack of time.

i really don't want to commit untested code, even though they are trivial changes. maybe someone else can apply them. it probably doesn't matter much, as i don't think many people will go though the ordeal of tapping the serial port on this thing.

excuse me for not finding time to reply earlier.

and again, thank you very much for your help.

Reply:
First of all, I also appreciate your efforts and contributions to OpenWrt.

Yeah, I'm sorry for my (even late) nitpicks. I agree that it is rather a cosmetic issue. It would be nice if you can test the change, but it is just a small style-only change after all. I think just a simple build test would suffice in this case.

For the label naming scheme, I sometimes feel that having address in the label make things complicated (e.g. similar devices sharing dtsi but having different addresses), but I don't think using array index is way better in that sense (e.g. what if the location is swapped?). And do we actually need iface name in the label? I would name it something like `macaddr_wlan5g` if I was the first who introduced nvmem-cells to mac address, but these labels were generated by a conversion script and I tend to follow conventions unless there's a good attractive reason. Because if we once accept a non-conventional format then another guy will just copy and paste it and I don't like things getting disorganized. Of course, all these are just my personal opinion.

I'm willing to merge this after the requested change, and I can even amend it myself if you provide me the MAC address list so that I can add it in the commit message. Or you can choose to wait for another core dev to merge this in this state.

Thanks for your patience and endeavor again.

Reply:
hi @mans0n

i suppose everything is taken care of:

https://github.com/openwrt/openwrt/compare/e31ec46b19d9933e6423e6a443311b67e1300cd4..dd52c04402bed06033f33105c15ee74e054f266e#diff-cb281a4e74086520b1a8bcf5b60c2923ad3988298218315ae442c953b0bf48ba

...plus the commit message

i took the router offline and got a full ubi backup of the deployed configuration and then tested this:
wlan adapters work and all adapters have their correct MACs. (fortunately i didn't need the serial.)

i also updated and tested the 22.03 branch which i hadn't in a while and everything checks out (https://github.com/openwrt/openwrt/pull/9658).

thanks for your help!